In memory without temporary file

vmx opened the pull request:
This code shows how to do things in memory without a temporary file.

The point here is that the `write_padded()` call is still needed. This call is not about doing the padding with zeros, but it is manipulating the data itself. For more information see https://github.com/filecoin-project/rust-fil-proofs/blob/0f9bd279c678a2f41bdad8bf9755a5125654574f/filecoin-proofs/src/fr32.rs.

So this code just creates a huge in-memory vector holding the data which would normally be written to the temporary file.

My guess is (I haven't checked) that the process could probably be streamed, so that you don't need to allocate for the full file size. But that surely would be a major change to the code (and perhaps even not easily possible).

---

The last commit is the actual change, all priory ones were just improvements of the original code.

vmx committed:
fix: update Cargo.lock

vmx committed:
refactor: use Rust 2018 edition

That makes importing external crates easier.

vmx committed:
refactor: run `cargo fmt` for nicer formatting

vmx committed:
refactor: improve code

vmx committed:
feat: make it work without temp file

Keep everything in memory instead of writing it to a temporary file.

mikeal commented:
awesome!

I’m expecting roughly 2x the target file size to be consumed by memory. Could we get this down if we kept the files below 1GB rather than slightly over?

Can we test real quick with this CAR file https://dumbo-stage3-fast-ai-nlp.s3-us-west-2.amazonaws.com/fast-ai-nlp-4.car that is 962MB ;)

vmx committed:
fix: lower the memory consumption

Make the vector's capacity just a bit bigger than it needs to be, so
that it isn't grown automatically. This keeps the memory consumption
lower.

rvagg commented:
Merged, also added a monitoring script that works on linux and monitors memory usage and tmpdir usage, it turns out there's even more temporary file work going on here. I added an option to revert to using the filecoin_proofs method so you can switch between them to test. Using the original 1G test file:

* Using the filecoin_proofs method yields a peak disk usage of 6,144 Mb and a peak memory usage of 4 Mb.
* Running the same using the storage_proofs method, avoiding the initial padded tmpfile and managing the padded version in memory yields a peak disk usage 4,096 Mb of and a peak memory usage of 2,053 Mb.

So a double allocation of the padded size on disk is still going on somewhere. Will hunt this down more.

rvagg commented:
Using the smaller file (961 Mb) it doesn't have to round up to 2G:

* original method: Peak disk 3072 Mb, Peak mem 4 Mb
* direct (the one in this PR) method: Peak disk 2012 Mb, Peak mem 1029 Mb

half the time too, 8 seconds instead of 16.

mikeal commented:
you’re still seeing disc usage that high with the in-memory changes?

rvagg commented:
yes, still hunting down what it is but it's in storage_proofs somewhere, I know that for sure

rvagg commented:
Yeah, this comes from https://github.com/filecoin-project/merkle_light

It needs a caching store for all of the intermediate pieces. rust-fil-proofs hardwires in a `DiskStore` (a `VecStore` is apparently also possible for in-memory): https://github.com/filecoin-project/rust-fil-proofs/blob/0f9bd279c678a2f41bdad8bf9755a5125654574f/storage-proofs/src/merkle.rs#L22-L23

I guess this makes sense since it's doing a merkle proof in chunks of 32 bytes (afaik) with an arity of 2 (afaik) so that's a ton of chunks at a ton of levels for a 1G piece, double the size in cache seems like it might be reasonable.

This might be immovable, we can offload the padded form into memory like in this PR, we could also probably hack in a `VecStore` in a custom fork of the proofs or re-implementation of the bits that call merkle_light, but then we're taking on a lot of custom code but we're still stuck with roughly 3 times the amount of memory required to process a single piece.

mikeal commented:
> This might be immovable, we can offload the padded form into memory like in this PR, we could also probably hack in a VecStore in a custom fork of the proofs or re-implementation of the bits that call merkle_light, but then we're taking on a lot of custom code but we're still stuck with roughly 3 times the amount of memory required to process a single piece.

As long as we can get a CAR file that is under 1GB to consume less than 3GB of memory and little to no disc, we’re fine, so let’s avoid forking the code beyond what is necessary to hit that target.
